iceoryx_utils/test/moduletests/test_resizeable_lockfree_queue.cpp

```diff
@@ -0,0 +1,402 @@
+    }
+}
+
+// note this is one of the most general cases and necessary to test:
+// decreasing the capacity starting with a partially filled queue and checking whether the last values
+// remain (and the others are removed)
+TYPED_TEST(ResizeableLockFreeQueueTest, DecreaseCapacityOfAPartiallyFilledQueue)
+{
+    auto& q = this->queue;
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+    iox::cxx::vector<int, MAX_CAP> removedElements;
+
+    uint64_t cap = MAX_CAP / 2;
+
+    EXPECT_TRUE(q.setCapacity(cap));
+    EXPECT_EQ(q.capacity(), cap);
+
+    uint64_t element = 0;
+    while (q.tryPush(element++))
+        ;
+
+    EXPECT_EQ(q.capacity(), cap);
+    EXPECT_EQ(q.size(), cap);
+
+    auto cap2 = cap + MAX_CAP / 4; // roughly 3 quarters of max (integer division)
+    EXPECT_TRUE(q.setCapacity(cap2));
+
+    EXPECT_EQ(q.capacity(), cap2);
+    EXPECT_EQ(q.size(), cap);
+
+    auto cap3 = cap2 - cap; // roughly a quarter of max
+
+    EXPECT_TRUE(q.setCapacity(cap3, removedElements));
+    EXPECT_EQ(q.capacity(), cap3);
+    EXPECT_EQ(q.size(), cap3);
+
+    // cap3 elements remain, the first cap - cap3 elements are removed
+
+    // were the least recent elements removed?
+    EXPECT_EQ(removedElements.size(), cap - cap3);
+    element = 0;
+    for (auto& removedElement : removedElements)
+    {
+        EXPECT_EQ(removedElement, element++);
+    }
+
+    // are the remaining elements correct? (i.e. we did not remove too many elements)
+    for (element = cap - cap3; element < cap; ++element)
+    {
+        auto result = q.pop();
+        ASSERT_TRUE(result.has_value());
+        EXPECT_EQ(result.value(), element);
+    }
+
+    // refill to verify the capacity can really be used
+
+    element = 0;
+    while (q.tryPush(element++))
+        ;
+
+    for (element = 0; element < cap3; ++element)
+    {
+        auto result = q.pop();
+        ASSERT_TRUE(result.has_value());
+        EXPECT_EQ(result.value(), element);
+    }
+}
+
+TYPED_TEST(ResizeableLockFreeQueueTest, DecreaseCapacityOfAPartiallyFilledQueueWithRemoveHandler)
+{
+    using element_t = typename TestFixture::Queue::element_t;
+    auto& q = this->queue;
+    constexpr auto MAX_CAP = TestFixture::Queue::MAX_CAPACITY;
+
+    iox::cxx::vector<int, MAX_CAP> removedElements;
+    auto removeHandler = [&](const element_t& value) { removedElements.push_back(std::move(value)); };
+
+    uint64_t cap = MAX_CAP / 2;
+
+    EXPECT_TRUE(q.setCapacity(cap));
+    EXPECT_EQ(q.capacity(), cap);
+
+    uint64_t element = 0;
+    while (q.tryPush(element++))
+        ;
+
+    EXPECT_EQ(q.capacity(), cap);
+    EXPECT_EQ(q.size(), cap);
+
+    auto cap2 = cap + MAX_CAP / 4; // roughly 3 quarters of max (integer division)
+    EXPECT_TRUE(q.setCapacity(cap2));
+
+    EXPECT_EQ(q.capacity(), cap2);
+    EXPECT_EQ(q.size(), cap);
+
+    auto cap3 = cap2 - cap; // roughly a quarter of max
+
+    EXPECT_TRUE(q.setCapacity(cap3, removeHandler));
```

Comment:
One reson for the complexity is to setup the most general state: 0<size<current capacity and then decrease the capacity. Sharing code in combination with typed tests is difficult since the presence of a removehandler requires checking the elements removed I will see what is possible here.

Reply:
I added even more tests to test every case with the removeHandler as well (as in principle it is a different method, even when the plain one currently calls the one with the removeHandler). Now, to do this together with typed tests leads to too much effort and more convoluted setups, but I did not fully try (not worth it from the readability point of view in my opinion). 

It could be an idea to setup a method to redirect (either with or without handler) in the test fixture. But it gets more complicated, the handler version needs to do additional checks on the container and so on. While it is more code now, it should be more readable than a potential generalization, on indiviudal test level at least.

TLDR: decided against making it even more generic but added missing tests with the handler...